File: 2024/with_elixir/lib/day1.ex
defmodule Day1 do
  def part1(content) do
    {lefts, rights} = from_strings_to_pairs(content)

    Enum.zip(
      Enum.sort(lefts),
      Enum.sort(rights)
    )
    |> Enum.map(fn {left, right} -> abs(left - right) end)
    |> Enum.sum()
  end

  def part2(content) do
    {lefts, rights} = from_strings_to_pairs(content)
    occurrences = Enum.frequencies(rights)

    lefts
    |> Enum.map(fn left -> left * Map.get(occurrences, left, 0) end)
    |> Enum.sum()
  end

  defp from_strings_to_pairs(content) do
    content
    |> Enum.map(fn x -> String.split(x) end)
    |> List.flatten()
    |> Enum.map(fn x -> String.to_integer(x) end)
    |> Enum.chunk_every(2)
    |> Enum.map(fn [a, b] -> {a, b} end)
    |> Enum.unzip()
  end
end

File: 2024/with_python/src/day1.py
from collections import Counter


def part1(content: list[str]) -> int:
    left_numbers, right_numbers = zip(*(map(int, entry.split()) for entry in content))
    return sum(
        abs(left - right)
        for left, right in zip(sorted(left_numbers), sorted(right_numbers))
    )


def part2(content: list[str]) -> int:
    left_numbers, right_numbers = zip(*(map(int, entry.split()) for entry in content))

    return sum(left * Counter(right_numbers)[left] for left in left_numbers)
